### PR TITLE
Migrate reentrancyguard to cairo2

### LINKED ISSUE #656: Migrate ReentrancyGuard to Cairo 2

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes #656 

Depends on #661 

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/tests/mocks/reentrancy_mock.cairo [hunk 1]
```diff
@@ -1,93 +1,92 @@
...
-
-    // Dispatchers
-    use super::IReentrancyGuardedDispatcher;
-    use super::IReentrancyGuardedDispatcherTrait;
     use openzeppelin::tests::mocks::reentrancy_attacker_mock::IAttackerDispatcher;
     use openzeppelin::tests::mocks::reentrancy_attacker_mock::IAttackerDispatcherTrait;
-
-    // Other
-    use option::OptionTrait;
     use starknet::ContractAddress;
```
Comment (ericnordelo) [inline src/openzeppelin/tests/mocks/reentrancy_mock.cairo:22]: We've been putting oz imports below starknet imports, I think we should be consistent with the order.

File: src/openzeppelin/tests/mocks/reentrancy_mock.cairo [hunk 2]
```diff
@@ -1,93 +1,92 @@
...
     struct Storage {
         counter: felt252
     }
 
-    #[view]
-    fn current_count() -> felt252 {
-        counter::read()
-    }
+    #[generate_trait]
+    impl StorageImpl of StorageTrait {
```
Comment (ericnordelo) [inline src/openzeppelin/tests/mocks/reentrancy_mock.cairo:33]:
```suggestion
    impl InternalImpl of InternalTrait {
```

File: src/openzeppelin/tests/mocks/reentrancy_attacker_mock.cairo [hunk 3]
```diff
@@ -1,21 +1,23 @@
-#[abi]
-trait IAttacker {
-    fn call_sender();
+#[starknet::interface]
```
Comment (ericnordelo) [inline src/openzeppelin/tests/mocks/reentrancy_attacker_mock.cairo:1]: If we are not using this Dispatcher, I think we can remove this interface, and implement the method directly as external in the mock for simplicity.

**Review by ericnordelo — COMMENTED** [review]
Looking good Andrew! Left some comments.

File: src/openzeppelin/tests/mocks/reentrancy_mock.cairo [hunk 4]
```diff
@@ -1,93 +1,92 @@
...
-    #[external]
-    fn count_local_recursive(n: felt252);
-    #[external]
-    fn count_external_recursive(n: felt252);
-    #[external]
-    fn count_and_call(attacker: ContractAddress);
-    #[external]
-    fn count();
+#[starknet::interface]
+trait IReentrancyMock<TContractState> {
```
Comment (ericnordelo) [inline src/openzeppelin/tests/mocks/reentrancy_mock.cairo:9]:
```suggestion
trait IReentrancyMock<TState> {
```

I think we should use this in mock as well for consistency.

Comment (andrew-fleming) [reply src/openzeppelin/tests/mocks/reentrancy_mock.cairo:9]: Good call, will update

**Review by ericnordelo — COMMENTED** [review]
LGTM!

**Review by ericnordelo — APPROVED** [review]

**Pull request merged**